ngraph/core/reference/include/ngraph/runtime/reference/max_pool.hpp

```diff
@@ -110,6 +110,340 @@ void max_pool(const T* arg,
     }
     NGRAPH_SUPPRESS_DEPRECATED_END
 }
+
+namespace {
+void validate_max_pool_kernel_params(const size_t dims,
+                                     const Shape& kernel,
+                                     const Strides& kernel_strides,
+                                     const Strides& kernel_dilations,
+                                     const Shape& pads_begin,
+                                     const Shape& pads_end) {
+    NGRAPH_CHECK(kernel.size() == dims && kernel_strides.size() == dims && kernel_dilations.size() == dims &&
+                     pads_begin.size() == dims && pads_end.size() == dims,
+                 "One of the MaxPool params does not match the ",
+                 dims,
+                 "D implementation.\nkernel=",
+                 kernel,
+                 "\nkernel_strides=",
+                 kernel_strides,
+                 "\nkernel_dilations=",
+                 kernel_dilations,
+                 "\npads_begin=",
+                 pads_begin,
+                 "\npads_end=",
+                 pads_end);
+}
+
+template <typename T>
+struct Coord : public std::vector<T> {
+    Coord(const Shape& pads_begin) {
+        std::vector<T>::reserve(pads_begin.size());
+        for (const auto axis_padding : pads_begin) {
+            std::vector<T>::push_back(0 - axis_padding);
+        }
+    }
+
+    Coord(std::initializer_list<T>&& values) : std::vector<T>{std::move(values)} {}
+};
+
+bool elem_in_padding_area(const Coord<int>& kernel_position,
+                          const Coord<size_t>& kernel_offset,
+                          const Shape& data_shape) {
+    for (size_t dim = 0; dim < data_shape.size() - 2; ++dim) {
```

Comment:
Need an assert that data_shape.size() >= 2. Otherwise we may have an infinite loop

Reply:
Or even better solution is to change the code the following way:
```cpp
for (size_t dim = 0; dim + 2 < data_shape.size(); ++dim) {
```